### ISSUE TITLE
Should version numbers be included in vocabulary namespaces

### ISSUE THREAD

**Issue by markdstafford** [OData/vocabularies#1]
The [OData core vocabulary](http://docs.oasis-open.org/odata/odata/v4.0/os/vocabularies/Org.OData.Core.V1.xml) includes the version number in the vocabulary namespace in addition to the URL. This seems redundant and would require OData services that reference the vocabulary to keep namespace and URL in sync.

The proposed convention for community vocabularies stored here would be to include the version in the file name but not the vocabulary namespace.

**Comment by ralfhandl** [comment]
We want to compatibly evolve vocabularies according to the [Model Versioning rules](http://docs.oasis-open.org/odata/odata/v4.0/errata02/os/complete/part1-protocol/odata-v4.0-errata02-os-part1-protocol-complete.html#_Toc406398209), and [adding a term to a vocabulary](https://issues.oasis-open.org/browse/ODATA-769) will be considered a compatible change.

Yet we foresee the need to make incompatible changes some time in the future, e.g. getting rid of terms that have become obsolete. Which is why we included the version number from the start to indicate that there may be different versions in the future.

**Comment by markdstafford** [comment]
@ralfhandl - I get the need for versioning from the beginning. I think what I'm objecting to is the duplication of versioning. If I have the version in the file name, that means the version is in the URL. Why do I then also need the version in the namespace of my vocabulary?

Looking at the issue from a different angle, including the version in both places requires consumers to make two corresponding changes. Let's use the proposed display vocabulary as an example. Imagine that we make a breaking change to the core vocabulary and so increment its version number to V2. Here's the changes I have to make to my dependency:

``` xml
<?xml version="1.0" ?>
<edmx:Edmx Version="4.0" xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx">
  <!--I need to change the reference URI to include the new version number.-->
  <edmx:Reference Uri="http://docs.oasis-open.org/odata/odata/v4.0/os/vocabularies/Org.OData.Core.V2.xml">
    <!--I need to change the alias - or worse, every reference to the that vocabulary - to include the new version number.-->
    <edmx:Include Alias="Core" Namespace="Org.OData.Core.V2"/>
  </edmx:Reference>
  <edmx:DataServices>
    <Schema Namespace="OData.Community.Display" xmlns="http://docs.oasis-open.org/odata/ns/edm">
      <Term Name="DisplayableValue" Type="Edm.String">
        <Annotation Term="Core.Description">
          <String>Provides a value designed primarily for display. For additional details see https://github.com/OData/vocabularies/blob/master/OData.Community.Display.V1.md</String>
        </Annotation>
      </Term>
    </Schema>
  </edmx:DataServices>
</edmx:Edmx>
```

We obviously need to keep the version number in the file name - otherwise we have a fundamental problem versioning shared vocabularies. But I would propose that the version number is unnecessary in the namespace of the vocabulary.

Thoughts?

**Comment by ralfhandl** [comment]
The namespace identifies the vocabulary. The reference uri just tells you where you can fetch a copy of the vocabulary.

The Core vocabulary is identified by its namespace Org.OData.Core.V1.
You can fetch it from either of these urls:
http://docs.oasis-open.org/odata/odata/v4.0/errata02/os/complete/vocabularies/Org.OData.Core.V1.xml
http://docs.oasis-open.org/odata/odata/v4.0/errata01/os/complete/vocabularies/Org.OData.Core.V1.xml
http://docs.oasis-open.org/odata/odata/v4.0/os/vocabularies/Org.OData.Core.V1.xml

I can also fetch it from http://xxxx/sap/opu/odata/IWFND/CATALOGSERVICE;v=2/Vocabularies(TechnicalName='%2FIWBEP%2FVOC_CORE',Version='0001',SAP__Origin='LOCAL')/$value

No way to reliably determine from the url which version of this vocabulary you will get. Only the namespace can reliably express that.

**Comment by markdstafford** [comment]
I might be missing something, so thanks for bearing with me as I work through this.

I just don't understand how having the version in both places is helpful. AFAICT a URL can only contain one vocabulary, right? So if there really is a dependency on an exact version of a vocabulary, both the URL and the namespace must be right. Let me try to think through this by walking through the things that could happen with the display vocabulary.

Starting state:
- The "qualified" vocabulary is hosted at `~/qualified.v1.xml`
- The "unqualified" vocabulary is hosted at `~/unqualified.v1.xml`
- The "qualified" vocabulary namespace is `qualified.v1`
- The "unqualified" vocabulary namespace is `unqualified`
- Dependent application is using 4 of the 5 terms in the v1 vocabulary and does not have an in-memory representation of the vocabulary (i.e., it must do an HTTP GET for the vocabulary at runtime)

| Change | "Qualified" Result | "Unqualified" Result |
| --- | --- | --- |
| Vocabulary is moved | The "qualified" vocabulary is now hosted at `~/qualified.new.xml`. Dependent breaks because `GET ~/qualified.v1.xml` returns `404`. **_Mitigation: Dependent changes reference URI.**_ | The "unqualified" vocabulary is now hosted at `~/unqualified.new.xml`. Dependent breaks because `GET ~/unqualified.v1.xml` returns `404`. **_Mitigation: Dependent changes reference URI.**_ |
| Vocabulary is versioned by copying to new URL | The "qualified" vocabulary now has two versions, v1 hosted at `~/qualified.v1.xml` with a namespace of `qualified.v1` and v2 hosted at `~/qualified.v2.xml` with a namespace of `qualified.v2`. No effect on dependent. **_Mitigation: Not necessary.**_ | The "unqualified" vocabulary now has two versions, v1 hosted at `~/unqualified.v1.xml` with a namespace of `unqualified.v1` and v2 hosted at `~/unqualified.v2.xml` with a namespace of `unqualified.v2`. No effect on dependent. **_Mitigation: Not necessary.**_ |
| <span class="datahighlight">Contents of vocabulary are versioned in-place, unused term is removed | The "qualified" vocabulary is still hosted at `~/qualified.v1.xml`. The namespace is now `qualified.v2`. Dependent breaks because no terms can be resolved. **_Mitigation: Dependent changes namespace.**_ | The "unqualified" vocabulary is still hosted at `~/unqualified.v1.xml`. The namespace is still `qualified`. Dependent does not break. **_Mitigation: Not needed.**_</span> |
| Contents of vocabulary are versioned in-place, used term is removed | Dependent breaks because no terms can be resolved. **_Mitigation: Not possible.**_ | Dependent breaks because one term cannot be resolved. **_Mitigation: Not possible.**_ |

Most of the scenarios wind up with the same result. However, there's at least one scenario that has a better result if the version is **_not**_ included in the namespace. We also know that it's more work for dependents to update references to the vocabulary since they have to change the version in two places. Is there a scenario I missed where having the version in both places would have yielded the better result?

**Comment by ralfhandl** [comment]
In most (all?) of our use cases the client has built-in knowledge of the vocabulary and only needs the namespace to decide whether it knows the terms. E.g. UI controls interpret Measures.ISOCurrency for displayinh/formatting/editing of monetary amounts, and the client-side model uses Capabilities terms to decide which requests it can fire. 
Terms are "interface" or "protocol" definitions. Design-time tools read and interpret the term definitions, Data clients just interpret the annotations for terms they know.

**Comment by markdstafford** [comment]
I'm not sure I buy into the built-in argument but I get your point.

@dingeins had (IMO) a better point as we discussed the other day that convinced me we need the version number in both places. Consider a REST API that wants to annotate resources with the `Contact` vocabulary. There are three versions of the `Contact` vocabulary and the server has clients that know about all three versions. The REST API would likely annotate resources using all three versions of the vocabulary, which would create a conflict if both the URL and namespace are not versioned.

I will add a version to Display's namespace.

**Comment by markdstafford** [comment]
Fixed in commit 639f05ba6a8d708ecfdc39207057cb8abe6c2bf6. We will include version in vocabulary namespaces.
